android-webrtc/URLVideoCapturer/urlcapturer/src/main/java/com/amazonaws/kinesisvideo/webrtc/urlvideocapturer/URLVideoCapturer.java

```diff
@@ -0,0 +1,145 @@
+ */
+public class URLVideoCapturer implements VideoCapturer, VideoSink {
+    private Uri uri;
+    private String[] options;
+    private String aspectRatio;
+    private SurfaceTextureHelper surfaceTextureHelper;
+    private Context context;
+    private CapturerObserver capturerObserver;
+
+    /**
+     * Public constructor accepting the stream url and libVLC configuration.
+     *
+     * @param uri         Media stream uri.
+     * @param options     List of options to pass on to libVLC.
+     * @param aspectRatio Aspect ratio to configure for libVLC.
+     */
+    public URLVideoCapturer(Uri uri, String[] options, String aspectRatio) {
+        if (uri == null) {
+            throw new IllegalArgumentException("URI is null.");
+        }
+
+        this.uri = uri;
+        this.options = options;
+        this.aspectRatio = aspectRatio;
+    }
+
+    /**
+     * Initializes this object by setting up the SurfaceTextureHelper.
+     *
+     * @param surfaceTextureHelper The SurfaceTextureHelper instance to use.
+     * @param context              Android Context object to be used by libVLC.
+     * @param capturerObserver     An instance of CapturerObserver to be notified about video frames.
+     */
+    @Override
+    public void initialize(SurfaceTextureHelper surfaceTextureHelper, Context context, CapturerObserver capturerObserver) {
+        this.surfaceTextureHelper = surfaceTextureHelper;
+        this.context = context;
+        this.capturerObserver = capturerObserver;
+
+        surfaceTextureHelper.startListening(this);
+    }
+
+    /**
+     * Callback to be used by the SurfaceTextureHelper when a video frame is ready.
+     *
+     * @param videoFrame VideoFrame from the SurfaceTextureHelper.
+     */
+    @Override
+    public void onFrame(VideoFrame videoFrame) {
+        capturerObserver.onFrameCaptured(videoFrame);
+    }
+
+    /**
+     * Begins the stream capture at specified resolution.
+     *
+     * @param width     Width of the frame.
+     * @param height    Height of the frame.
+     * @param framerate This parameter is ignored.
+     */
+    @Override
+    public void startCapture(int width, int height, int framerate) {
+        capturerObserver.onCapturerStarted(true);
+        surfaceTextureHelper.setTextureSize(width, height);
+
+        // Use libVLC to play the stream onto the texture.
+        LibVLC libVlc = options == null ? new LibVLC(context) : new LibVLC(context, new ArrayList<>(Arrays.asList(options)));
+
+        MediaPlayer mediaPlayer = new MediaPlayer(libVlc);
+        IVLCVout vOut = mediaPlayer.getVLCVout();
+        vOut.setWindowSize(width, height);
+        vOut.setVideoSurface(surfaceTextureHelper.getSurfaceTexture());
+        vOut.attachViews();
+
+        Media videoMedia = new Media(libVlc, uri);
+        mediaPlayer.setMedia(videoMedia);
+        mediaPlayer.setAspectRatio(aspectRatio == null ? "4:3" : aspectRatio);
```

Comment:
Do we want to allow null? https://javadoc.io/doc/org.videolan.android/libvlc-all/3.2.6/org/videolan/libvlc/MediaPlayer.html#setAspectRatio-java.lang.String- says:
> Parameters:
aspect - new video aspect-ratio or NULL to reset to default

Also just curious to know what happens when we pass in empty string, their docs doesn't say what happens.

Reply:
Thanks for this feedback. I checked with null and empty strings. The player still renders fine in some default format. So, I removed the unnecessary check on our side.